nixos/modules/programs/dconf.nix

```diff
@@ -1,55 +1,182 @@
 { config, lib, pkgs, ... }:
 
-with lib;
-
 let
   cfg = config.programs.dconf;
-  cfgDir = pkgs.symlinkJoin {
-    name = "dconf-system-config";
-    paths = map (x: "${x}/etc/dconf") cfg.packages;
-    postBuild = ''
-      mkdir -p $out/profile
-      mkdir -p $out/db
-    '' + (
-      concatStringsSep "\n" (
-        mapAttrsToList (
-          name: path: ''
-            ln -s ${path} $out/profile/${name}
-          ''
-        ) cfg.profiles
+
+  # Compile keyfiles to dconf DB
+  compileDconfDb = dir: pkgs.runCommand "dconf-db"
+    {
+      nativeBuildInputs = [ (lib.getBin pkgs.dconf) ];
+    } "dconf compile $out ${dir}";
+
+  # Recursively lock values in settings
+  recursiveLock =
+    settings: locks:
+    (lib.mapAttrsRecursive
+      (path: value:
+      lib.optionalAttrs value
+        (lib.mapAttrsRecursiveCond
+          (value': !lib.gvariant.isGVariant value')
+          (path': value': true)
+          (lib.foldl'
+            (root: name: if lib.hasAttr "${name}" root then root."${name}" else
+            throw "${lib.concatStringsSep "." (lib.tail path)} doesn't exist in settings.")
+            { v = settings; }
+            path)
+        )
       )
-    ) + ''
-      ${pkgs.dconf}/bin/dconf update $out/db
-    '';
+      { v = locks; }).v;
+
+  # Generate dconf DB from attrset and keyfiles
```

Comment:
Then it will be possible to say that `val` has the type of `dconfDatabase` submodule.

Reply:
Do you mean adding it to the comment?